edgelet/aziot-edged/src/image_gc.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use edgelet_core::{Module, ModuleRuntime};
+use std::collections::HashMap;
+use std::fs;
+use std::io::Write;
+use std::time::{Duration, UNIX_EPOCH};
+
+use crate::error::Error as EdgedError;
+
+// TODO: fix file name with correct path
+const FILE_NAME: &str = "abc.txt";
+
+pub(crate) async fn run_until_shutdown(
+    settings: edgelet_settings::docker::Settings,
+    runtime: &edgelet_docker::DockerModuleRuntime<http_common::Connector>,
+) -> Result<edgelet_core::WatchdogAction, EdgedError> {
+    // Run the GC once a day while waiting for any running task
+    let gc_period = std::time::Duration::from_secs(86400);
+
+    let mut gc_timer = tokio::time::interval(gc_period);
+    gc_timer.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+    log::info!("Starting MIGC...");
+
+    loop {
+        if let Err(err) = garbage_collector(&settings, runtime).await {
+            return Err(EdgedError::new(format!("Error in MIGC: {}", err)));
+        }
+    }
+}
+
+async fn garbage_collector(
+    _settings: &edgelet_settings::docker::Settings,
+    runtime: &edgelet_docker::DockerModuleRuntime<http_common::Connector>,
+) -> Result<(), EdgedError> {
+    log::info!("MIGC job starting daily run");
+
+    // Step 1: read MIGC persistence file into in-mem map
+    // this map now contains all images deployed to the device (through an IoT Edge deployment)
+    let image_map = get_images_with_timestamp().map_err(|e| e)?;
+
+    /* ============================== */
+
+    // Step 2: track images associated with extant containers
+
+    // first get list of containers on the device, running or otherwise
+    let running_modules = ModuleRuntime::list_with_details(runtime).await.unwrap();
+
+    /* ============================== */
+
+    // Step 3: process maps
+    let (image_map, carry_over) = process_state(image_map, running_modules);
+
+    /* ============================== */
+
+    // Step 4: delete images
+
+    for key in image_map.keys() {
+        let result = ModuleRuntime::remove(runtime, key).await;
```

Comment:
There's an overload: https://github.com/gauravIoTEdge/iotedge/blob/master/edgelet/edgelet-docker/src/runtime.rs#L119

My understanding is that deletion can be by - in this case - image name or image ID: https://docs.docker.com/engine/api/v1.41/#tag/Image/operation/ImageDelete

If I'm missing something here, then please call it out.

Reply:
Ooh, that's interesting. Yes, it looks like `DockerModuleRuntime` does implement `fn remove` for the `ModuleRegistry` trait, so we'd just need to make sure we're using that instead of the `ModuleRuntime` trait.